0000-api-versions.md

```diff
@@ -0,0 +1,50 @@
+# Meta
+[meta]: #meta
+- Name: API Versions
+- Start Date: 2019-08-05
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Begin versioning the various APIs that the spec describes. Then require these versions be specified in related TOML files, under an `api` field.
+
+# Motivation
+[motivation]: #motivation
+
+To enable logic and/or validation based on `api`. For instance, we could keep users from attempting to use a combination of incompatible buildpacks and lifecycle.
+
+# What it is
+[what-it-is]: #what-it-is
+
+There would be two pieces to this:
+
+1. A reference to the current versions of all APIs placed in the spec. Currently, we would version the Buildpack API (buildpack-to-lifecycle contract) and the Platform API (lifecycle-to-platform contract).
+2. An `api` field in TOML files for entities that are dependent on a specific version of an API. For instance, `buildpack.toml` would have an `api` to represent the version of the buildpack API to which the buildpack conforms.
+
+Example `api` field for `buildpack.toml`:
+
+```toml
+api = "<major>.<minor>"
+
+[buildpack]
+id = "<string>"
+# ...
+```
+
+The format of a version would be `<major>.<minor>` or `<major>` (with a `minor` of `0` implied). A change to the `major` version of an API would indicate a non-backwards-compatible change, while a change to the `minor` version would indicate availability of one or more opt-in features.
+
+A **missing or empty** `api` key would indicate `1.0` (the current version of the Buildpack API and Platform API, as of this writing). The next breaking change to either of the current APIs would bump that API version to `2.0`.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Where in the spec should we document current API versions?
```

Comment:
Does it make sense to just put it at the top of the respective `*.md` files?

Reply:
What's interesting about this question is that the next API version we would support doesn't match the spec exactly at any commit. Example: we could put API version 2.0 at the top of the `buildpack.md` but that would seem to imply support for `store.toml` which isn't yet implemented. We could describe what 2.0 entails at the top of the spec and say `2.1` will add store.toml?

Reply:
I wonder if we just need to version the spec in a branch (or tag) that's just accurate for that version. It's why @jkutner and I were talking about just 1 version of the API. It'd make that logistically easier.

Reply:
Regardless of how we version the spec, there are two contracts that need separate versions. Otherwise, if we make a breaking change for platform maintainers that doesn't affect buildpacks, we would be forced to push a new major version on buildpack authors for no reason.

Reply:
@sclevine i wouldn't say it's for no reason; it's so only one version has to be managed. given how infrequently we will have breaking changes (especially after finalizing the spec) I suspect it won't be that big of a burden. That said, I'm fine if we decide on separate versions.